File: ios/modules/NativePingStorage/NativePingStorageModule.swift
//
//  NativePingStorageModule.swift
//  RNSampleApp
//
//  Created by Thais Damasceno on 2025-09-18.
//

import Foundation
import PingStorage
import React
import PingJourney
import PingLogger
import PingOrchestrate

@objc(NativePingStorageModule)
class NativePingStorageModule: NSObject {
  private var storage: (any Storage<String>)?
  
  enum CacheStrategy {
    case noCache
    case cache
    case cacheOnFailure
  }
  
  
  private func parseCacheStrategy(from value: String?) -> CacheStrategy {
    print("NativePingStorageModule: Parsing cache strategy from value: \(String(describing: value))")
    switch value?.uppercased() {
    case "CACHE":
      return .cache
    case "CACHE_ON_FAILURE":
      return .cacheOnFailure
    default:
      return .noCache
    }
  }
  
  @objc
  func configure(_ config: [String: Any]) {
    print("NativePingStorageModule: configure called with config: \(config)")
    do {
      let type = config["type"] as? String ?? "encrypted"
      let keyAlias = (config["keyAlias"] as? String) ?? "com.pingidentity.rnsampleapp.keyalias"
      let cacheStrategy = parseCacheStrategy(from: config["cacheStrategy"] as? String)
      print("NativePingStorageModule: type=\(type), keyAlias=\(keyAlias), cacheStrategy=\(cacheStrategy)")
      let base: any Storage<String>
      switch type {
      case "memory":
        print("NativePingStorageModule: Using MemoryStorage")
        base = MemoryStorage<String>()
      case "encrypted", "datastore":
        print("NativePingStorageModule: Using KeychainStorage")
        base = KeychainStorage<String>(
          account: keyAlias,
          encryptor: NoEncryptor()
        )
      default:
        print("NativePingStorageModule: Unknown type, defaulting to MemoryStorage")
        base = MemoryStorage<String>()
      }
      
      switch cacheStrategy {
      case .noCache:
        print("NativePingStorageModule: No cache strategy")
        storage = base
      case .cache, .cacheOnFailure:
        print("NativePingStorageModule: Using StorageDelegate with cacheable true")
        storage = StorageDelegate(delegate: base, cacheable: true)
      }
      print("NativePingStorageModule: Storage configured successfully")
    } catch {
      print("NativePingStorageModule: Error configuring storage: \(error)")
    }
  }
  
  @objc
  func save(_ item: [String: Any], completion: @escaping(NSNumber?, NSError?) -> Void) {
    print("NativePingStorageModule: save called with item: \(item)")
    Task {
      do {
        let data = try JSONSerialization.data(withJSONObject: item, options: [])
        let jsonString = String(data: data, encoding: .utf8) ?? "{}"
        print("NativePingStorageModule: Saving jsonString: \(jsonString)")
        try await storage?.save(item: jsonString)
        print("NativePingStorageModule: Save successful")
        completion(NSNumber(value: true), nil)
      } catch {
        print("NativePingStorageModule: Error saving item: \(error)")
        completion(nil, error as NSError)
      }
    }
  }
  
  @objc
  func get(_ completion: @escaping(Any?, NSError?) -> Void) {
    print("NativePingStorageModule: get called")
    Task {
      do {
        if let json = try await storage?.get() {
          print("NativePingStorageModule: Retrieved json: \(json)")
          if let data = json.data(using: .utf8) {
            let item = try JSONSerialization.jsonObject(with: data, options: [])
            print("NativePingStorageModule: Returning item: \(item)")
            completion(item, nil)
          } else {
            print("NativePingStorageModule: Could not convert json to data")
            completion(nil, nil)
          }
        } else {
          print("NativePingStorageModule: No data found in storage")
          completion(nil, nil)
        }
      } catch {
        print("NativePingStorageModule: Error getting item: \(error)")
        completion(nil, error as NSError)
      }
    }
  }
  
  @objc
  func remove(_ completion: @escaping(NSNumber?, NSError?) -> Void) {
    print("NativePingStorageModule: remove called")
    Task {
      do {
        try await storage?.delete()
        print("NativePingStorageModule: Remove successful")
        completion(NSNumber(value: true), nil)
      } catch {
        print("NativePingStorageModule: Error removing item: \(error)")
        completion(nil, error as NSError)
      }
    }
  }
  
  // MARK: - Journey Methods
  
  private var journey: Journey?
  private var lastNode: Node?
  
  // Helper to convert Node → JS object
  private func serializeNode(_ node: Node) -> NSDictionary {
    var dict: [String: Any] = ["id": UUID().uuidString]
    
    switch node {
    case let continueNode as ContinueNode:
      dict["type"] = "ContinueNode"
      dict["callbacks"] = continueNode.callbacks.map { cb in
        [
          "type": String(describing: type(of: cb)),
          "prompt": (cb as? TextOutputCallback)?.message ?? "",
          "value": NSNull()
        ]
      }
    case let errorNode as ErrorNode:
      dict["type"] = "ErrorNode"
      dict["message"] = errorNode.message
    case let failureNode as FailureNode:
      dict["type"] = "FailureNode"
      dict["cause"] = failureNode.cause.localizedDescription
    case let successNode as SuccessNode:
      dict["type"] = "SuccessNode"
      dict["session"] = successNode.session
    default:
      dict["type"] = "Unknown"
    }
    
    return dict as NSDictionary
  }
  
  
  // MARK: - Configure Journey (already working)
  @objc(configureJourney:completion:)
  func configureJourney(_ config: NSDictionary,
                        completion: @escaping (NSNumber?, NSError?) -> Void) {
    print("NativePingStorageModule: configureJourney called with \(config)")
    
    guard let serverUrl = config["serverUrl"] as? String else {
      let error = NSError(domain: "NativePingStorageModule",
                          code: 400,
                          userInfo: [NSLocalizedDescriptionKey: "Missing serverUrl"])
      completion(nil, error)
      return
    }
    
    let realm = config["realm"] as? String
    let cookieName = config["cookie"] as? String
    let clientId = config["clientId"] as? String
    let discoveryEndpoint = config["discoveryEndpoint"] as? String
    let redirectUri = config["redirectUri"] as? String
    let scopes = config["scopes"] as? [String] ?? ["openid", "email", "profile"]
    
    // Create and configure Journey
    self.journey = Journey.createJourney { journeyConfig in
      journeyConfig.serverUrl = serverUrl
      if let realm = realm { journeyConfig.realm = realm }
      if let cookie = cookieName { journeyConfig.cookie = cookie }
      journeyConfig.timeout = 30
      journeyConfig.logger = LogManager.standard
      
      if let clientId = clientId,
         let discoveryEndpoint = discoveryEndpoint,
         let redirectUri = redirectUri {
        journeyConfig.module(PingJourney.OidcModule.config) { oidcValue in
          oidcValue.clientId = clientId
          oidcValue.discoveryEndpoint = discoveryEndpoint
          oidcValue.redirectUri = redirectUri
          oidcValue.scopes = Set(scopes)
        }
      }
    }
    
    if self.journey != nil {
      print("✅ Journey configured successfully for \(serverUrl)")
      completion(true, nil)
    } else {
      let error = NSError(domain: "NativePingStorageModule",
                          code: 500,
                          userInfo: [NSLocalizedDescriptionKey: "Failed to configure Journey"])
      completion(nil, error)
    }
  }
  
  
  // MARK: - Start Journey
  @objc(start:options:completion:)
  func start(_ journeyName: String,
             options: NSDictionary?,
             completion: @escaping (NSDictionary?, NSError?) -> Void) {
    print("NativePingStorageModule: start called for \(journeyName) with options \(String(describing: options))")
    
    guard let journey = self.journey else {
      let error = NSError(domain: "NativePingStorageModule",
                          code: 500,
                          userInfo: [NSLocalizedDescriptionKey: "Journey not configured"])
      completion(nil, error)
      return
    }
    
    let forceAuth = options?["forceAuth"] as? Bool ?? false
    let noSession = options?["noSession"] as? Bool ?? false
    
    Task {
      let node = await journey.start(journeyName) { opts in
        opts.forceAuth = forceAuth
        opts.noSession = noSession
      }
      self.lastNode = node
      completion(self.serializeNode(node), nil)
    }
  }
  
  
  // MARK: - Next
  @objc(next:input:completion:)
  func next(_ nodeId: NSString, // TODO : Revisit if we need nodeId or keep rely on internal lastNode
            input: NSDictionary,
            completion: @escaping (NSDictionary?, NSError?) -> Void) {
    print("NativePingStorageModule: next called with nodeId \(nodeId) and input \(input)")
    
    guard let current = self.lastNode as? ContinueNode else {
      let error = NSError(domain: "NativePingStorageModule",
                          code: 404,
                          userInfo: [NSLocalizedDescriptionKey: "No active ContinueNode"])
      completion(nil, error)
      return
    }
    
    if let callbacks = input["callbacks"] as? [[String: Any]] {
      for cb in callbacks {
        if let type = cb["type"] as? String {
          switch type {
          case "NameCallback":
            if let nameCB = current.callbacks.first(where: { $0 is NameCallback }) as? NameCallback {
              nameCB.name = cb["value"] as? String ?? ""
            }
          case "PasswordCallback":
            if let passCB = current.callbacks.first(where: { $0 is PasswordCallback }) as? PasswordCallback {
              passCB.password = cb["value"] as? String ?? ""
            }
          default:
            break
          }
        }
      }
    }
    
    Task {
      let nextNode = await current.next()
      self.lastNode = nextNode
      completion(self.serializeNode(nextNode), nil)
    }
  }
  
  
  // MARK: - Resume
  @objc(resume:completion:)
  func resume(_ uri: NSString,
              completion: @escaping (NSDictionary?, NSError?) -> Void) {
    print("NativePingStorageModule: resume called with uri \(uri)")
    
    guard let journey = self.journey else {
      let error = NSError(
        domain: "NativePingStorageModule",
        code: 400,
        userInfo: [NSLocalizedDescriptionKey: "Journey not configured"]
      )
      completion(nil, error)
      return
    }
    
    // Convert string → URL
    guard let resumeUrl = URL(string: uri as String) else {
      let error = NSError(
        domain: "NativePingStorageModule",
        code: 422,
        userInfo: [NSLocalizedDescriptionKey: "Invalid resume URI format"]
      )
      completion(nil, error)
      return
    }
    
    // Execute the resume call
    Task {
      let resumedNode = await journey.resume(resumeUrl)
      self.lastNode = resumedNode
      completion(self.serializeNode(resumedNode), nil)
    }
  }
  
  
  // MARK: - Get Session
  @objc(getSession:)
  func getSession(_ completion: @escaping (NSDictionary?, NSError?) -> Void) {
    print("NativePingStorageModule: getSession called")
    
    guard let journey = self.journey else {
      let error = NSError(
        domain: "NativePingStorageModule",
        code: 400,
        userInfo: [NSLocalizedDescriptionKey: "Journey not configured"]
      )
      completion(nil, error)
      return
    }
    
    Task {
      let user = await journey.journeyUser()
      
      // Fetch OIDC user info (Result<UserInfo, OidcError>)
      let userInfoResult = await user?.userinfo(cache: false)
      var userInfoDict: [String: Any] = [:]
      
      if let result = userInfoResult {
        switch result {
        case .success(let info):
          // Convert Swift dictionary to [String: Any]
          userInfoDict = info.mapValues { $0 }
          print("✅ UserInfo received: \(userInfoDict)")
        case .failure(let error):
          print("⚠️ UserInfo fetch failed: \(error.localizedDescription)")
        }
      }
      
      // Fetch token
      let tokenResult = await user?.token()
      
      switch tokenResult {
      case .success(let token):
        var dict: [String: Any] = [
          "accessToken": token.accessToken,
          "refreshToken": token.refreshToken ?? "",
          "expiresIn": token.expiresIn
        ]
        
        if !userInfoDict.isEmpty {
          dict["userInfo"] = userInfoDict
        }
        
        print("✅ Returning session object with user info")
        completion(dict as NSDictionary, nil)
        
      case .failure(let error):
        let nsError = NSError(
          domain: "NativePingStorageModule",
          code: 401,
          userInfo: [NSLocalizedDescriptionKey: error.localizedDescription]
        )
        completion(nil, nsError)
        
      case .none:
        print("⚠️ No token available")
        completion(nil, nil)
      }
    }
  }
  
  
  // MARK: - Logout
  @objc(logout:)
  func logout(_ completion: @escaping (NSNumber?, NSError?) -> Void) {
    print("NativePingStorageModule: logout called")
    
    guard let journey = self.journey else {
      completion(false, NSError(domain: "NativePingStorageModule",
                                code: 400,
                                userInfo: [NSLocalizedDescriptionKey: "Journey not configured"]))
      return
    }
    
    Task {
      let user = await journey.journeyUser()
      await user?.logout()
      completion(true, nil)
    }
  }
  
}

File: android/app/src/main/java/com/rnsampleapp/modules/storage/NativePingStorageModule.kt
package com.rnsampleapp.modules.storage

import android.content.Context
import android.util.Log
import androidx.datastore.core.DataStore
import androidx.datastore.core.DataStoreFactory
import com.facebook.fbreact.specs.NativePingStorageSpec
import com.facebook.react.bridge.*
import com.pingidentity.journey.*
import com.pingidentity.journey.callback.NameCallback
import com.pingidentity.journey.callback.PasswordCallback
import com.pingidentity.journey.callback.TextInputCallback
import com.pingidentity.journey.callback.TextOutputCallback
import com.pingidentity.journey.module.Oidc
import com.pingidentity.journey.module.Session
import com.pingidentity.journey.module.SessionConfig
import com.pingidentity.journey.module.RequestUrl
import com.pingidentity.journey.plugin.callbacks
import com.pingidentity.journey.SSOToken
import com.pingidentity.journey.journey
import com.pingidentity.journey.options
import com.pingidentity.orchestrate.Node
import com.pingidentity.orchestrate.ContinueNode
import com.pingidentity.orchestrate.ErrorNode
import com.pingidentity.orchestrate.SuccessNode
import com.pingidentity.orchestrate.FailureNode
import com.pingidentity.orchestrate.Session
import com.pingidentity.orchestrate.module.Cookie
import com.pingidentity.orchestrate.EmptySession
import com.pingidentity.orchestrate.Module
import com.pingidentity.oidc.*
import com.pingidentity.oidc.module.*
import com.pingidentity.oidc.exception.AuthorizeException
import com.pingidentity.storage.*
import com.pingidentity.utils.Result
import com.pingidentity.utils.Result.Failure
import com.pingidentity.utils.Result.Success
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.JsonArray
import kotlinx.coroutines.*
import org.json.JSONObject
import java.io.File
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor

fun createStringDataStore(context: Context, fileName: String): DataStore<String?> {
    return DataStoreFactory.create(
        serializer = DataToJsonSerializer(),
        scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    ) {
        File(context.filesDir, "datastore/$fileName")
    }
}

class NativePingStorageModule(context: ReactApplicationContext) :
    NativePingStorageSpec(context) {

    private val scope = CoroutineScope(Dispatchers.IO)
    private var storage: Storage<String>? = null
    // Journey SDK instance
    private var journey: Journey? = null
    // The most recent node from the Journey SDK
    private var lastNode: Node? = null

    override fun getName() = "NativePingStorage"

    override fun configure(config: ReadableMap, promise: Promise) {
        scope.launch {
            try {
                Log.d("NativePingStorage", "Configuring storage with: $config")

                val type = config.getString("type") ?: "encrypted"
                val fileName = config.getString("fileName") ?: "secure_prefs"
                val cacheStrategy = when (config.getString("cacheStrategy")) {
                    "CACHE" -> CacheStrategy.CACHE
                    "CACHE_ON_FAILURE" -> CacheStrategy.CACHE_ON_FAILURE
                    else -> CacheStrategy.NO_CACHE
                }

                storage = when (type) {
                    "memory" -> {
                        Log.d("NativePingStorage", "Using MemoryStorage")
                        MemoryStorage<String>()
                    }
                    "datastore" -> {
                        Log.d("NativePingStorage", "Using DataStoreStorage with fileName=$fileName")
                        val dataStore = createStringDataStore(reactApplicationContext, fileName)
                        DataStoreStorage(
                            dataStore = dataStore,
                            cacheStrategy = cacheStrategy
                        )
                    }
                    else -> {
                        val keyAlias = config.getString("keyAlias") ?: "defaultKey"
                        val strongBox = config.getBoolean("strongBoxPreferred")
                        Log.d(
                            "NativePingStorage",
                            "Using EncryptedDataStoreStorage with fileName=$fileName, keyAlias=$keyAlias, strongBox=$strongBox"
                        )
                        EncryptedDataStoreStorage<String> {
                            this.fileName = fileName
                            this.cacheStrategy = cacheStrategy
                            this.keyAlias = keyAlias
                            this.strongBoxPreferred = strongBox
                        }
                    }
                }

                Log.d("NativePingStorage", "Storage configured successfully ✅")
                promise.resolve(true)
            } catch (e: Exception) {
                Log.e("NativePingStorage", "Error configuring storage", e)
                promise.reject("CONFIG_ERROR", e)
            }
        }
    }

    override fun save(item: ReadableMap, promise: Promise) {
        scope.launch {
            try {
                val jsonString = JSONObject(item.toHashMap()).toString()
                Log.d("NativePingStorage", "Saving item: $jsonString")
                storage?.save(jsonString)
                Log.d("NativePingStorage", "Item saved ✅")
                promise.resolve(true)
            } catch (e: Exception) {
                Log.e("NativePingStorage", "Error saving item", e)
                promise.reject("SAVE_ERROR", e)
            }
        }
    }

    override fun get(promise: Promise) {
        scope.launch {
            try {
                Log.d("NativePingStorage", "Fetching item from storage...")
                val jsonString = storage?.get()
                if (jsonString != null) {
                    Log.d("NativePingStorage", "Retrieved raw JSON: $jsonString")
                    val json = JSONObject(jsonString)
                    val map = Arguments.createMap()
                    json.keys().forEach { key ->
                        map.putString(key, json.getString(key))
                    }
                    Log.d("NativePingStorage", "Returning map: $map")
                    promise.resolve(map)
                } else {
                    Log.d("NativePingStorage", "No item found")
                    promise.resolve(null)
                }
            } catch (e: Exception) {
                Log.e("NativePingStorage", "Error fetching item", e)
                promise.reject("GET_ERROR", e)
            }
        }
    }

    override fun remove(promise: Promise) {
        scope.launch {
            try {
                Log.d("NativePingStorage", "Deleting item from storage...")
                storage?.delete()
                Log.d("NativePingStorage", "Item deleted ✅")
                promise.resolve(true)
            } catch (e: Exception) {
                Log.e("NativePingStorage", "Error deleting item", e)
                promise.reject("DELETE_ERROR", e)
            }
        }
    }


    // ===== Journey Methods =====

    /**
     * Helper function to convert a Journey SDK Node into a WritableMap for React Native.
     */
    private fun serializeNode(node: Node): WritableMap {

        val map = Arguments.createMap()
        map.putString("id", node.hashCode().toString())

        when (node) {
            is ContinueNode -> {
                Log.d("NativePingStorage", "Serializing ContinueNode with ${node.callbacks.size} callbacks")
                map.putString("type", "ContinueNode")
                val callbacksArray = Arguments.createArray()
                node.callbacks.forEach { cb ->
                    val callbackMap = Arguments.createMap()
                    callbackMap.putString("type", cb::class.java.simpleName)
                    // Add specific properties based on callback type
                    when (cb) {
                        is TextOutputCallback -> {
                            callbackMap.putString("message", cb.message)
                            callbackMap.putString("prompt", cb.message)
                        }
                        is TextInputCallback -> {
                            callbackMap.putString("prompt", cb.prompt)
                            callbackMap.putString("value", cb.text ?: "")
                        }
                        is PasswordCallback -> {
                            callbackMap.putString("prompt", cb.prompt)
                            callbackMap.putString("value", "") // Don't expose password
                        }
                        is NameCallback -> {
                            callbackMap.putString("prompt", cb.prompt)
                            callbackMap.putString("value", cb.name ?: "")
                        }
                        else -> {
                            // For unknown callback types, try to get prompt if available
                            callbackMap.putString("prompt", "")
                            callbackMap.putString("value", "")
                        }
                    }
                    callbacksArray.pushMap(callbackMap)
                }
                map.putArray("callbacks", callbacksArray)
            }
            is ErrorNode -> {
                map.putString("type", "ErrorNode")
                map.putString("message", node.message)
            }
            is SuccessNode -> {
                map.putString("type", "SuccessNode")
                // The session should be automatically established by the SDK when SuccessNode is received
                // The user can retrieve session data via getSession() method
                Log.d("NativePingStorage", "Serialized SuccessNode - session established, data available via getSession()")
            }
            is FailureNode -> {
                map.putString("type", "FailureNode")
                map.putString("message", node.cause.message ?: node.cause.toString())
                Log.d("NativePingStorage", "Serialized FailureNode with message: ${node.cause.message ?: node.cause.toString()}")
            }
            else -> {
                map.putString("type", "UnknownNode")
            }
        }
        Log.d("NativePingStorage", "Serialized node map: ${map.toString()}")
        return map
    }
    
    override fun configureJourney(config: ReadableMap, promise: Promise) {
        try {
            Log.d("NativePingStorage", "configureJourney called with: $config")

            if (!config.hasKey("serverUrl")) {
                promise.reject("CONFIG_ERROR", "Missing required parameter: serverUrl")
                return
            }
            val serverUrl = config.getString("serverUrl")!!

            // Optional parameters
            val realm = config.getString("realm")
            val cookieName = config.getString("cookie")
            val clientId = config.getString("clientId")
            val discoveryEndpoint = config.getString("discoveryEndpoint")
            val redirectUri = config.getString("redirectUri")
            val scopesArray = config.getArray("scopes")?.toArrayList()?.mapNotNull { it.toString() }

            // Initialize the Journey SDK 
            Log.d("NativePingStorage", "Initializing Journey SDK...")
            
            this.journey = Journey {
                this.timeout = 30000 // Network request timeout in milliseconds (default: 15000ms)
                this.serverUrl = serverUrl
                realm?.let { this.realm = it }
                cookieName?.let { 
                    this.cookie = it

                }
                
                // Configure OIDC module 
                if (clientId != null && discoveryEndpoint != null && redirectUri != null) {
                     Log.d("NativePingStorage", "Configuring OIDC module...")
                    this.module(Oidc) {
                        this.clientId = clientId
                        this.discoveryEndpoint = discoveryEndpoint
                        this.redirectUri = redirectUri
                        this.scopes = scopesArray?.toMutableSet() ?: mutableSetOf("openid", "email", "address", "profile", "phone")
                    }
                    Log.d("NativePingStorage", "OIDC module configured successfully")
                } else {
                    Log.d("NativePingStorage", "OIDC module NOT configured - missing parameters")
                }
            }
            if(this.journey != null) {
                Log.d("NativePingStorage", "Journey SDK initialized successfully")
                promise.resolve(true)
            } else {
                Log.d("NativePingStorage", "Journey SDK initialization FAILED")
            }
        } catch (e: Exception) {
            Log.e("NativePingStorage", "Error configuring Journey", e)
            promise.reject("CONFIG_ERROR", "Failed to configure Journey: ${e.message}", e)
        }
    }


    override fun start(journeyName: String, options: ReadableMap?, promise: Promise) {
        Log.d("NativePingStorage", "Start called for journey: '$journeyName' with options: $options")

        val journeyInstance = this.journey
        if (journeyInstance == null) {
            promise.reject("NOT_CONFIGURED", "Journey SDK has not been configured. Call configureJourney() first.")
            return
        }

        val forceAuth = options?.getBoolean("forceAuth") ?: false
        val noSession = options?.getBoolean("noSession") ?: false

        scope.launch {
            try {
                Log.d("NativePingStorage", "Starting journey with forceAuth=$forceAuth, noSession=$noSession")
                val node = journeyInstance.start(journeyName) {
                    this.forceAuth = forceAuth
                    this.noSession = noSession
                }
                this@NativePingStorageModule.lastNode = node
                val serialized = serializeNode(node)
                promise.resolve(serialized)
            } catch (e: Exception) {
                Log.e("NativePingStorage", "Exception in start: ${e.message}", e)
                promise.reject("START_ERROR", "Failed to start journey '$journeyName': ${e.message}", e)
            }
        }
    }

    override fun next(nodeId: String, input: ReadableMap, promise: Promise) {
        Log.d("NativePingStorage", "next called for nodeId: $nodeId with input: $input")

        val journeyInstance = this.journey
        if (journeyInstance == null) {
            promise.reject("NOT_CONFIGURED", "Journey SDK has not been configured. Call configureJourney() first.")
            return
        }

        val currentNode = this.lastNode
        if (currentNode == null) {
            promise.reject("NO_ACTIVE_JOURNEY", "There is no active journey in progress. Call start() first.")
            return
        }

        if (currentNode !is ContinueNode) {
            promise.reject("INVALID_STATE", "The current journey node is not a ContinueNode and does not accept input.")
            return
        }

        // Apply input from the ReadableMap to the corresponding callbacks
        try {
            val inputMap = input.toHashMap()
            
            if (inputMap.containsKey("callbacks")) {
                val callbacksArray = inputMap["callbacks"] as? ArrayList<*>
                Log.d("NativePingStorage", "Found callbacks array with ${callbacksArray?.size} items")
                
                callbacksArray?.forEach { callbackData ->
                    val callbackMap = callbackData as? HashMap<*, *>
                    val type = callbackMap?.get("type") as? String
                    val value = callbackMap?.get("value") as? String ?: ""
                    
                    when (type) {
                        "NameCallback" -> {
                            val nameCB = currentNode.callbacks.firstOrNull { it is NameCallback } as? NameCallback
                            nameCB?.let {
                                it.name = value
                            }
                        }
                        "PasswordCallback" -> {
                            val passCB = currentNode.callbacks.firstOrNull { it is PasswordCallback } as? PasswordCallback
                            passCB?.let {
                                it.password = value
                            }
                        }
                        "TextInputCallback" -> {
                            val textCB = currentNode.callbacks.firstOrNull { it is TextInputCallback } as? TextInputCallback
                            textCB?.let {
                                it.text = value
                            }
                        }
                        else -> {
                            Log.w("NativePingStorage", "Unhandled callback type in next: $type. No input applied.")
                        }
                    }
                }
            } else {
                Log.w("NativePingStorage", "No 'callbacks' key found in input. Available keys: ${inputMap.keys}")
            }
        } catch (e: Exception) {
            promise.reject("INPUT_ERROR", "Failed to set callback values: ${e.message}", e)
            return
        }

        scope.launch {
            try {
                val nextNode = currentNode.next()
                this@NativePingStorageModule.lastNode = nextNode
                promise.resolve(serializeNode(nextNode))
            } catch (e: Exception) {
                Log.e("NativePingStorage", "Exception in next: ${e.message}", e)
                promise.reject("NEXT_ERROR", "Failed to proceed to next node: ${e.message}", e)
            }
        }
    }

    override fun resume(uri: String, promise: Promise) {
        Log.d("NativePingStorage", "resume called with uri: $uri")
        
        val journeyInstance = this.journey
        if (journeyInstance == null) {
            promise.reject("NOT_CONFIGURED", "Journey SDK has not been configured. Call configureJourney() first.")
            return
        }
        
        scope.launch {
            try {
                val resumedNode = journeyInstance.resume(android.net.Uri.parse(uri))
                this@NativePingStorageModule.lastNode = resumedNode
                val serialized = serializeNode(resumedNode)
                promise.resolve(serialized)
            } catch (e: Exception) {
                Log.e("NativePingStorage", "Exception in resume: ${e.message}", e)
                promise.reject("RESUME_ERROR", "Failed to resume journey with URI '$uri': ${e.message}", e)
            }
        }
    }

    override fun getSession(promise: Promise) {
        
        val journeyInstance = this.journey
        if (journeyInstance == null) {
            promise.reject("NOT_CONFIGURED", "Journey SDK has not been configured. Call configureJourney() first.")
            return
        }
        
        scope.launch {
            try {
                Log.d("NativePingStorage", "Starting getting user session...")

                val user = journeyInstance.user()

                if (user == null) {
                    Log.d("NativePingStorage", "Get session failed - No user available")
                    promise.resolve(null)
                    return@launch
                }
                
                Log.d("NativePingStorage", "User object retrieved, fetching session data...")
                
                // Fetch token()  - returns Result<Token, OidcError>
                when (val tokenResult = user.token()) {
                    is Result.Success -> {
                        val token = tokenResult.value
                        
                        // Fetch userinfo() only if token was successful - returns Result<JsonObject, OidcError>
                        var userInfoMap: WritableMap? = null
                        when (val result = user.userinfo(false)) {
                            is Result.Failure -> {
                                Log.w("NativePingStorage", "Error fetching user info: ${result.value}")
                                userInfoMap = null
                            }
                            is Result.Success -> {
                                val userInfoJson = result.value
                                userInfoMap = Arguments.createMap()
                                userInfoJson.forEach { (key, value) ->
                                    when (value) {
                                        is JsonPrimitive -> {
                                            when {
                                                value.isString -> userInfoMap.putString(key, value.content)
                                                else -> userInfoMap.putString(key, value.toString())
                                            }
                                        }
                                        is JsonArray -> {
                                            userInfoMap.putString(key, value.toString()) // Convert arrays to string
                                        }
                                        is JsonObject -> {
                                            userInfoMap.putString(key, value.toString()) // Convert nested objects to string
                                        }
                                        else -> {
                                            userInfoMap.putString(key, value.toString())
                                        }
                                    }
                                }
                                Log.d("NativePingStorage", "User info fetched successfully")
                            }
                        }

                        val resultMap = Arguments.createMap()
                        resultMap.putString("accessToken", token.accessToken)
                        resultMap.putString("refreshToken", token.refreshToken ?: "")
                        resultMap.putLong("expiresIn", token.expiresIn)
                        
                        if (userInfoMap != null) {
                            resultMap.putMap("userInfo", userInfoMap)
                        }

                        promise.resolve(resultMap)
                    }
                    is Result.Failure -> {
                        Log.e("NativePingStorage", "Error fetching token: ${tokenResult.value}")
                        // No valid session - reject with clear error message
                        promise.reject("NO_SESSION", "No active session. Please authenticate first by completing the journey flow.")
                        return@launch
                    }
                }
            } catch (e: Exception) {
                Log.e("NativePingStorage", "Error in getSession", e)
                promise.reject("GET_SESSION_ERROR", "Failed to get session: ${e.message}", e)
            }
        }
    }

    override fun logout(promise: Promise) {
        
        val journeyInstance = this.journey
        if (journeyInstance == null) {
            promise.reject("NOT_CONFIGURED", "Journey SDK has not been configured. Call configureJourney() first.")
            return
        }
        
        scope.launch {
            try {
                val user = journeyInstance.user()
                user?.logout()
                Log.d("NativePingStorage", "User logged out successfully")
                promise.resolve(true)
            } catch (e: Exception) {
                Log.e("NativePingStorage", "Error during logout", e)
                promise.reject("LOGOUT_ERROR", "Failed to logout: ${e.message}", e)
            }
        }
    }

}
